### ISSUE TITLE
Inference Issues as Grid Patterns

### ISSUE THREAD

**Issue by VC86** [Jyxarthur/flowsam#12]
Hello,

I have been trying to run inference with FlowI-SAM (using ViT-H weights) using:
- DAVIS2016, by rerunning the dataset annotations using your `flow/run_inference.py` script
- The evaluation script you provided, with arguments `--ckpt_path frame_level_flowisam_vith_train_on_oclrsyn_dvs16.pth --flow_encoder vit_h --flow_encoder_ckpt_path sam_vit_h_4b8939.pth --rgb_encoder_ckpt_path sam_vit_h_4b8939.pth --save_path /datasets/test`

However, I observe some unexpected patterns in the resulting segmentation masks, for example the image below, corresponding to the `blackswan` sample `00003.png` in the dataset with the usual `1, -1, 2, -2` flow pairs, yielding
![00003](https://github.com/user-attachments/assets/7188b65c-5326-48b0-8f64-5488fd52a9ac)

For reference, when inspecting the DAVIS 2016 dataset folder, the optical flow (e.g., gap 1) appears to be correctly saved
![00003](https://github.com/user-attachments/assets/b0a008f6-bd57-4305-8a24-e8293fb896f4)

The observed grid pattern-like artifacts seem to be appearing very often across this sequence as well as in other examples I tested. Is there any known workaround, or could it point to any error in the way I am using the `evaluation.py` script as I expect these artifacts would not appear in your results?

Thanks for your great work btw.

EDIT: I also tested `frame_level_flowpsam_vitbvith_train_on_oclrsyn_dvs16.pth`. It seems to be delivering less grid artefacts, but these are still present (e.g., in the `breakdance` sample) with for example the default settings of `evaluation.py` (i.e., all but those that _must_ be specified set to default values of the script).

**Comment by VC86** [comment]
Thank you @Jyxarthur for your feedback. I appreciate the tips on background/foreground regions handling and on making decisions based on the scores.

My question was more pertaining to why I see this type of "ringing artifacts" in the labels:
![image](https://github.com/user-attachments/assets/cfa11e88-fee6-4173-baf0-1a5dab99c3fc)
Are they caused by model output processing? I have observed such ringing patterns at the direct FlowSAM model output as well (i.e., in the raw mask logits); what could I do to mitigate them/get rid of them besides rule-based tricks (morphological operators, downsample/upsample, and the like)?

**Comment by Jyxarthur** [comment]
Thanks for the clarification. As noted in the previous response, this artifact is mainly due to minimal (unbalanced) loss applied to background regions, where the model is uncertain about the segmentation mask outputs (this often happens when flow is the only input). The pattern itself is suspected to be associated with grid-like positional encoding or ConvNet kernel, and the hard threshold would make this artifact more obvious.

I am afraid the simplest way to mitigate this issue is to adjust the thresholding on the logits (e.g. applying a lower threshold for the background masks). To better tackle this problem, there are several options to explicitly tune the model, e.g. by applying larger loss on background masks (as previously suggested) or applying a larger loss weight on the mask boundary.

**Comment by VC86** [comment]
Thanks, on a careful read of your first answer indeed all the elements were there, it was just not immediate for me to grasp these details. Alright, fair point, can close the issue.
